### PR TITLE
[Observable] chore: observable touchup

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Human Summary

- add `Observer#IsClosed()` to prevent redundant unsubscription
- simplify `channelObservable`
- update observable pkg README
- update pkg README template
- add `Observable#Next()` conveniece method to just get the next value

### Summary generated by Reviewpad on 23 Oct 23 16:25 UTC
This pull request includes the following changes:

1. In the file `docs/pkg/observable/README.md`, the package path has been updated from `pkg/observable` to `pocket/pkg/observable`.
2. In the file `docs/template/pkg/README.md`, the section "Architecture Diagrams" has been added to provide visual representations of the package's design and flow.
3. In the file `pkg/observable/channel/observable.go`, the `Next` method has been added to synchronously return the next value from the observable.
4. In the file `pkg/observable/channel/observable.go`, the `goPublish` function has been modified to use the `publishCh` field of the `channelObservable` struct.
5. In the file `pkg/observable/channel/observable_test.go`, new test cases have been added to test different scenarios of the `NotifyObservers` function.
6. In the file `pkg/observable/channel/observable_test.go`, the `UnsubscribeObservers` function now logs a warning when the observer is closed.
7. In the file `pkg/observable/channel/observer.go`, the `IsClosed` method has been added to check if the observer has been unsubscribed.
8. In the file `pkg/observable/interface.go`, the `Next` method has been added to the `Observable` interface to synchronously return the next value from the observable.
9. In the file `pkg/observable/interface.go`, the `IsClosed` method has been added to the `Observer` interface to check if the observer has been unsubscribed.

## Issue

- Follow-up to #31 

[Explain the reasoning for the PR in 1-2 sentences. Consider adding a link or a screenshot.]

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [x] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_test`
- [x] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

**Comment by bryanchriswhite** [comment]
Will need to merge main once #91 is in to pass CI.

File: docs/template/pkg/README.md [hunk 1]
```diff
@@ -1,6 +1,3 @@
-Certainly! I've added a section named "Architecture Diagrams" in the documentation template below:
-
-```markdown
 # Package [PackageName]
 
 > Brief one-liner or quote about what this package does.
```
Comment (Olshansk) [inline docs/template/pkg/README.md:3]: Do you mean to add TODOs here?

File: docs/template/pkg/README.md [hunk 2]
```diff
@@ -16,11 +13,21 @@ Provide a few sentences about the purpose and functionality of this package. Con
 
 Visual representations often make it easier to understand the design and flow of a package. Below are the architecture diagrams that explain the high-level structure and interactions in this package:
 
-![Architecture Overview](./path-to-diagram1.png)
+```mermaid
```
Comment (Olshansk) [inline docs/template/pkg/README.md:16]: Are these changes intended?

File: docs/template/pkg/README.md [hunk 3]
```diff
@@ -16,11 +13,21 @@ Provide a few sentences about the purpose and functionality of this package. Con
...
+---
+title: Architecture Overview
+---
+flowchart
+```
 
 > **Figure 1**: Brief description about what this diagram represents.
```
Comment (Olshansk) [inline docs/template/pkg/README.md:23]: Plaeholder?

File: pkg/observable/channel/observable.go [hunk 4]
```diff
@@ -63,6 +64,14 @@ func WithPublisher[V any](publishCh chan V) option[V] {
 	}
 }
 
+func (obsvbl *channelObservable[V]) Next(ctx context.Context) V {
```
Comment (Olshansk) [inline pkg/observable/channel/observable.go:67]: Can you add a godoc on what this is or why its needed?

File: pkg/observable/interface.go [hunk 5]
```diff
@@ -11,6 +11,7 @@ import "context"
 // notified of new values asynchronously.
 // It is analogous to a publisher in a "Fan-Out" system design.
 type Observable[V any] interface {
+	Next(context.Context) V
```
Comment (Olshansk) [inline pkg/observable/interface.go:14]: Either add a comment or what this does or add a TODO so we tend to it in the future.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: docs/template/pkg/README.md (hunk 2 above)
Comment (bryanchriswhite) [reply docs/template/pkg/README.md:16]: Yes, this file is intended to be a template for generating  go package READMEs with LLMs.

File: pkg/observable/channel/observable.go (hunk 4 above)
Comment (bryanchriswhite) [reply pkg/observable/channel/observable.go:68]: Added a godoc comment. It's not strictly necessary but will be convenient in forthcoming PRs.

File: pkg/observable/interface.go (hunk 5 above)
Comment (bryanchriswhite) [reply pkg/observable/interface.go:15]: Thanks for calling this out! :raised_hands: 

I added godoc comments to all these.

File: pkg/observable/channel/observable.go (hunk 4 above)
Comment (Olshansk) [reply pkg/observable/channel/observable.go:68]: `It's not strictly necessary but will be convenient in forthcoming PRs.`

I'm a bit confused as to when we should or shouldn't use it. #PUC when you have time, but not a blocker in this PR.

**Review by Olshansk — APPROVED** [review]
